stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -608,26 +630,43 @@ internal class ChannelControllerImpl(
 
         return if (online) {
             // upload attachments
-            logger.logI("Uploading attachments for message with id ${newMessage.id} and text ${newMessage.text}")
-            newMessage.attachments = newMessage.attachments.map {
-                var attachment: Attachment = it
-                if (it.upload != null) {
-                    val result = uploadAttachment(it, attachmentTransformer)
-                    attachment.uploadComplete = result.isSuccess
-                    if (result.isSuccess) {
-                        attachment = result.data()
-                    } else {
-                        attachment.uploadState = Attachment.UploadState.Failed(result.error())
-                        logger.logE("Failed to upload attachment for message")
+
+            if (hasAttachments) {
+                logger.logI("Uploading attachments for message with id ${newMessage.id} and text ${newMessage.text}")
+
+                newMessage.attachments = newMessage.attachments.mapIndexed { i, attach ->
+                    var attachment: Attachment = attach
+                    val attachmentProgress = attachmentProgressList[i]
+
+                    if (attachment.upload != null) {
+                        val result = uploadAttachment(
+                            attachment,
+                            attachmentTransformer,
+                            attachmentProgress.toProgressCallback()
+                        )
+
+                        if (result.isSuccess) {
```

Comment:
It was previously a messy method, new lines make it more and more complicated. it requires refactoring... 😢 
What if extract result of upload attachments to some `handleUploadResult(result: Result<Attachment>, progressTracker: ProgressTracker)?`

Reply:
Do you mean  the part inside `if (result.isSuccess) {`? This part of the code wasn't changed, but I can extract this to another method and extract some parts of `sendMessage` to another method.